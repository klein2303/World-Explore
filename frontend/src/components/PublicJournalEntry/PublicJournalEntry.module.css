.reviewCard {
    background-color: #fcf5ff;
    padding: 15px;
    border-radius: 8px;
    box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
    width: 100%;
    max-width: 350px;
    color: #424242;
    height: 220px;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    position: relative;
    box-sizing: border-box; /* Include padding and border in the element's total width and height */
}

.reviewTitle {
    font-size: 16px;
    font-weight: bold;
    margin-bottom: 5px;
}

.reviewDate {
    font-size: 16px;
    margin-bottom: 10px;
    opacity: 0.8;
}

.reviewRating {
    font-size: 16px;
}

.reviewInfo {
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding-bottom: 5px;
    font-weight: 500;
}

.user {
    display: flex;
    align-items: center;
    gap: 6px;
}

.reviewInfoBox {
    flex-grow: 1;
    overflow: hidden;
    text-overflow: ellipsis;
}

.reviewText {
    font-size: 16px;
    line-height: 1.4;
    overflow: hidden;
    height: 65px; /* Limit height to create uniform card size */
    -webkit-line-clamp: 2; /* Limit the number of lines */
    line-clamp: 2; /* Standard property for compatibility */
    -webkit-box-orient: vertical;
}

.readMoreButton {
    background-color: #424242;
    color: white;
    padding: 2px 5px;
    border: none;
    border-radius: 4px;
    cursor: pointer;
    font-size: 12px;
    margin-top: 10px;
}

.readMoreButton:hover {
    background-color: #797979;
}

.starIcon {
    margin-right: 6px;
}

.profileIcon {
    margin-right: 5px;
}

.reviewer {
    vertical-align: middle;
}

@media (max-width: 768px) {
    .reviewCard {
        max-width: 100%;
        min-width: 100%;
        flex-basis: 100%;
        height: auto;
    }
    .reviewText {
        height: auto;
        -webkit-line-clamp: unset;
        line-clamp: unset;
    }
}
