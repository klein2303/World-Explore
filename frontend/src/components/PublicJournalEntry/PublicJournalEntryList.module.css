.entriesGrid {
    display: flex;
    flex-direction: row;
    gap: 30px 85px;
    flex-wrap: wrap;
    justify-content: center;
    align-content: flex-start;
}

@media (max-width: 768px) {
    .entriesGrid {
        gap: 20px;
    }
}