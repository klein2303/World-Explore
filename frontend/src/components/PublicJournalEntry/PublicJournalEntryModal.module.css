.modalBackdrop {
    position: fixed;
    top: 0;
    left: 0;
    width: 100vw;
    height: 100vh;
    background: rgba(0, 0, 0, 0.6);
    display: flex;
    justify-content: center;
    align-items: center;
    z-index: 1000;
}

.user {
    display: flex;
    align-items: center;
    gap: 6px;
}

.modalContent {
    background: #fcf5ff;
    border-radius: 8px;
    padding: 20px;
    width: 90%;
    max-width: 600px;
    max-height: 80vh;
    overflow-y: auto;
    box-shadow: 0px 5px 15px rgba(0, 0, 0, 0.3);
    position: relative;
}

body[data-theme='dark'] .modalContent {
    background: #8A86A4;
}

.closeButton {
    position: absolute;
    top: 10px;
    right: 15px;
    background: none;
    border: none;
    font-size: 24px;
    color: #333;
    cursor: pointer;
}

body[data-theme='dark'] .closeButton {
    color: white;
}

.modalTitle {
    font-size: 24px;
    font-weight: bold;
    margin-bottom: 8px;
}

.modalDate {
    font-size: 14px;
    color: #777;
    margin-bottom: 10px;
}

body[data-theme='dark'] .modalDate {
    color: #dfdfdf;
}

.modalInfo {
    display: flex;
    gap: 20px;
    align-items: center;
    margin-bottom: 15px;
    justify-content: space-between;
}

.modalRating,
.modalReviewer {
    display: flex;
    align-items: center;
    gap: 6px;
}

.modalText {
    font-size: 16px;
    line-height: 1.5;
    color: #333;
    margin-top: 10px;
}

body[data-theme='dark'] .modalText {
    color: white;
}
