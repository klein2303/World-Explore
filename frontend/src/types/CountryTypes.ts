export type CountryType = {
    name: string;
    continent: string;
    capital: string;
    largest_city: string;
    currency: string;
    language: string;
    population: number;
    land_area: number;
    agriculture_area: string;
    forest_area: string;
    co2_emissions: number;
    image: string;
};
