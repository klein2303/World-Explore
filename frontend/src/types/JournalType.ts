export type JournalTypeRead = {
    country: string;
    reviews: readReviewType[];
};

export type JournalTypeWrite = {
    country: string;
    reviews: writeReviewType[];
};

export type readReviewType = {
    title: string;
    date: string;
    rating: number;
    text: string;
    public: boolean;
    journal: {
        profile: {
            username: string;
        };
    };
};

export type writeReviewType = {
    title: string;
    date: string;
    rating: number;
    text: string;
    public: boolean;
    journal: {
        profile: {
            email: string;
        };
    };
}