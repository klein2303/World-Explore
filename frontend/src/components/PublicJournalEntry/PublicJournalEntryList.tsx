import { JournalTypeRead } from "../../types/JournalType";
import PublicJournalEntry from "./PublicJournalEntry";
import styles from "./PublicJournalEntryList.module.css";

type PublicJournalEntriesProps = {
    journal: JournalTypeRead;
};

const PublicJournalEntryList = ({ journal }: PublicJournalEntriesProps) => {
    const { country, reviews } = journal;

    return (
        <>
            {/* Grid for journal entries */}
            <article
                className={styles.entriesGrid}
                role="list"
                aria-description={`Public journal entries for ${country}`}>
                {reviews.map((review) => (
                    <div key={review.title} role="listitem">
                        <PublicJournalEntry review={review} />
                    </div>
                ))}
            </article>
        </>
    );
};

export default PublicJournalEntryList;
