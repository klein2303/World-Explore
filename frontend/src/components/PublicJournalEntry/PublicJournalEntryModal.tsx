import { useEffect } from "react";
import styles from "./PublicJournalEntryModal.module.css";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faStar } from "@fortawesome/free-solid-svg-icons";
import { CgProfile } from "react-icons/cg";
import { reviewType } from "../../types/JournalType";

type JournalEntryModalProps = {
    review: reviewType;
    onClose: () => void;
};

const PublicJournalEntryModal = ({ review, onClose }: JournalEntryModalProps) => {
    // Close modal on "Escape" key press
    useEffect(() => {
        const handleKeyDown = (event: KeyboardEvent) => {
            if (event.key === "Escape") {
                onClose();
            }
        };
        document.addEventListener("keydown", handleKeyDown);
        return () => document.removeEventListener("keydown", handleKeyDown);
    }, [onClose]);

    return (
        <aside
            className={styles.modalBackdrop}
            onClick={onClose}
            role="dialog"
            aria-modal="true"
            aria-labelledby={`review-title-${review.title}`}
            aria-describedby={`review-text-${review.text}`}>
            <div className={styles.modalContent} onClick={(e) => e.stopPropagation()} tabIndex={-1}>
                {/* Review title, date and close button */}
                <header>
                    <button onClick={onClose} className={styles.closeButton} aria-label="Close modal">
                        &times;
                    </button>
                    <h3 id={`review-title-${review.title}`} className={styles.modalTitle}>
                        {review.title}
                    </h3>
                    <p className={styles.modalDate} aria-label={`Date range: ${review.date}`}>
                        {review.date}
                    </p>
                </header>

                {/* Review and reviewer info */}
                <section className={styles.modalInfo} aria-labelledby="modal-rating modal-reviewer">
                    <p
                        id="modal-rating"
                        className={styles.modalRating}
                        aria-label={`Rating: ${review.rating} out of 5 stars`}>
                        <FontAwesomeIcon icon={faStar} className={styles.starIcon} aria-hidden="true" />
                        {review.rating}/5
                    </p>
                    <p id="modal-reviewer" className={styles.modalReviewer} aria-label="Reviewer: Ola Nordmann">
                        <CgProfile className={styles.profileIcon} aria-hidden="true" />
                        Ola Nordmann
                    </p>
                </section>
                {/* Review text */}
                <article id={`review-text-${review.title}`} className={styles.modalText}>
                    {review.text}
                </article>
            </div>
        </aside>
    );
};

export default PublicJournalEntryModal;
