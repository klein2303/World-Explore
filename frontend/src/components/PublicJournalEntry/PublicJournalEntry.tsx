import { useState, useEffect } from "react";
import { reviewType } from "../../types/JournalType";
import styles from "./PublicJournalEntry.module.css";
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faStar } from '@fortawesome/free-solid-svg-icons';
import { CgProfile } from "react-icons/cg";
import PublicJournalEntryModal from "./PublicJournalEntryModal";

type PublicJournalEntryProps = {
    review: reviewType;
};

const PublicJournalEntry = ({ review }: PublicJournalEntryProps) => {
    const [isExpanded, setIsExpanded] = useState(false);
    const [isMobile, setIsMobile] = useState(window.innerWidth <= 768);
    const [showModal, setShowModal] = useState(false);

    const truncatedText = review.text.length > 100 ? review.text.slice(0, 100) + "..." : review.text;

    // Resize handler to check for screen size
    useEffect(() => {
        const handleResize = () => setIsMobile(window.innerWidth <= 768);
        window.addEventListener("resize", handleResize);
        return () => window.removeEventListener("resize", handleResize);
    }, []);

    const handleReadMore = () => {
        if (isMobile) {
            setIsExpanded(!isExpanded);
        } else {
            setShowModal(true);
        }
    };

    return (
        <>
            <div className={styles.reviewCard}>
                <h3 className={styles.reviewTitle}>{review.title}</h3>
                <p className={styles.reviewDate}>{review.date}</p>
                <section className={styles.reviewInfo}>
                    <p className={styles.reviewRating}>
                        <FontAwesomeIcon icon={faStar} className={styles.starIcon} />{review.rating}/5
                    </p>
                    <p className={styles.reviewer}>
                        <CgProfile className={styles.profileIcon} />
                        Ola Nordmann
                    </p>
                </section>
                <section className={styles.reviewInfoBox}>
                    <p className={styles.reviewText}>{isExpanded || showModal ? review.text : truncatedText}</p>
                    {review.text.length > 100 && (
                        <button className={styles.readMoreButton} onClick={handleReadMore}>
                            {isExpanded ? "Read Less" : "Read More"}
                        </button>
                    )}
                </section>
            </div>
            {showModal && !isMobile && (
                <PublicJournalEntryModal review={review} onClose={() => setShowModal(false)} />
            )}
        </>
    );
};

export default PublicJournalEntry;
