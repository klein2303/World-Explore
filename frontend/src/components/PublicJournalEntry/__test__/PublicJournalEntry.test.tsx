import { render, screen, fireEvent } from "@testing-library/react";
import { describe, it, expect, vi } from "vitest";
import PublicJournalEntry from "../PublicJournalEntry";
import { readReviewType } from "../../../types/JournalType";

// Mocking the modal
vi.mock("../PublicJournalEntryModal", () => ({
    default: ({ onClose }: { onClose: () => void }) => (
        <div data-testid="mock-modal">
            <button onClick={onClose} aria-label="Close modal">
                Close
            </button>
        </div>
    ),
}));

describe("PublicJournalEntry Component", () => {
    const mockReview: readReviewType = {
        title: "Amazing Experience",
        date: "2024-11-20",
        rating: 5,
        text: "This experience was absolutely wonderful and unforgettable. The journey was packed with excitement and beauty!",
        public: true,
        journal: {
            profile: {
                username: "traveler123",
            },
        },
    };

    beforeEach(() => {
        // Reset window size before each test
        window.innerWidth = 1024;
        fireEvent.resize(window);
    });

    it("renders correctly and matches snapshot", () => {
        const { asFragment } = render(<PublicJournalEntry review={mockReview} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it("displays truncated text and 'Read More' button on desktop", () => {
        render(<PublicJournalEntry review={mockReview} />);

        // Ensure truncated text is displayed
        expect(screen.getByText(/This experience was absolutely wonderful and unforgettable/)).toBeInTheDocument();
        expect(screen.queryByText(mockReview.text)).not.toBeInTheDocument();

        // Verify "Read More" button
        const readMoreButton = screen.getByRole("button", { name: "Expand review text" });
        expect(readMoreButton).toBeInTheDocument();
    });

    it("expands text on 'Read More' button click in mobile view", () => {
        window.innerWidth = 768; // Simulate mobile view
        fireEvent.resize(window);

        render(<PublicJournalEntry review={mockReview} />);

        const readMoreButton = screen.getByRole("button", { name: "Expand review text" });
        fireEvent.click(readMoreButton);

        // Verify full text is displayed
        expect(screen.getByText(mockReview.text)).toBeInTheDocument();

        // Verify "Read Less" button is now displayed
        expect(screen.getByRole("button", { name: "Collapse review text" })).toBeInTheDocument();
    });

    it("opens the modal when 'Read More' is clicked in desktop view", () => {
        render(<PublicJournalEntry review={mockReview} />);

        const readMoreButton = screen.getByRole("button", { name: "Expand review text" });
        fireEvent.click(readMoreButton);

        // Verify modal is displayed
        expect(screen.getByTestId("mock-modal")).toBeInTheDocument();
    });

    it("closes the modal when close button is clicked in the modal", () => {
        render(<PublicJournalEntry review={mockReview} />);

        const readMoreButton = screen.getByRole("button", { name: "Expand review text" });
        fireEvent.click(readMoreButton);

        const closeModalButton = screen.getByRole("button", { name: "Close modal" });
        fireEvent.click(closeModalButton);

        // Verify modal is no longer displayed
        expect(screen.queryByTestId("mock-modal")).not.toBeInTheDocument();
    });

    it("updates state on window resize", () => {
        const { rerender } = render(<PublicJournalEntry review={mockReview} />);

        // Verify initial desktop state (truncated text and no modal)
        const truncatedTextDesktop = screen.getByText(/This experience was absolutely wonderful and unforgettable/);
        expect(truncatedTextDesktop).toBeInTheDocument();
        expect(screen.queryByTestId("mock-modal")).not.toBeInTheDocument();

        // Simulate resizing to mobile
        window.innerWidth = 768;
        fireEvent.resize(window);
        rerender(<PublicJournalEntry review={mockReview} />);

        // Resize back to desktop
        window.innerWidth = 1024;
        fireEvent.resize(window);
        rerender(<PublicJournalEntry review={mockReview} />);

        // In desktop, click "Read More" to open modal
        const readMoreButtonDesktop = screen.getByRole("button", { name: "Expand review text" });
        fireEvent.click(readMoreButtonDesktop);

        // Verify modal is displayed
        expect(screen.getByTestId("mock-modal")).toBeInTheDocument();
    });
});