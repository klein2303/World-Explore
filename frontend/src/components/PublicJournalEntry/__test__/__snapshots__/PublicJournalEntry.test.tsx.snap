// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PublicJournalEntry Component > renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <article
    aria-labelledby="review-title-Amazing Experience"
    class="_reviewCard_1b1c0f"
    role="region"
  >
    <header>
      <h3
        class="_reviewTitle_1b1c0f"
        id="review-title-Amazing Experience"
      >
        Amazing Experience
      </h3>
      <p
        class="_reviewDate_1b1c0f"
      >
        2024-11-20
      </p>
    </header>
    <div
      class="_reviewInfo_1b1c0f"
    >
      <p
        aria-description="Rating: 5 out of 5 stars"
        class="_reviewRating_1b1c0f"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-star _starIcon_1b1c0f"
          data-icon="star"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 576 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M316.9 18C311.6 7 300.4 0 288.1 0s-23.4 7-28.8 18L195 150.3 51.4 171.5c-12 1.8-22 10.2-25.7 21.7s-.7 24.2 7.9 32.7L137.8 329 113.2 474.7c-2 12 3 24.2 12.9 31.3s23 8 33.8 2.3l128.3-68.5 128.3 68.5c10.8 5.7 23.9 4.9 33.8-2.3s14.9-19.3 12.9-31.3L438.5 329 542.7 225.9c8.6-8.5 11.7-21.2 7.9-32.7s-13.7-19.9-25.7-21.7L381.2 150.3 316.9 18z"
            fill="currentColor"
          />
        </svg>
        5/5
      </p>
      <p
        class="_user_1b1c0f"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 640 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224 256A128 128 0 1 0 224 0a128 128 0 1 0 0 256zm-45.7 48C79.8 304 0 383.8 0 482.3C0 498.7 13.3 512 29.7 512H322.8c-3.1-8.8-3.7-18.4-1.4-27.8l15-60.1c2.8-11.3 8.6-21.5 16.8-29.7l40.3-40.3c-32.1-31-75.7-50.1-123.9-50.1H178.3zm435.5-68.3c-15.6-15.6-40.9-15.6-56.6 0l-29.4 29.4 71 71 29.4-29.4c15.6-15.6 15.6-40.9 0-56.6l-14.4-14.4zM375.9 417c-4.1 4.1-7 9.2-8.4 14.9l-15 60.1c-1.4 5.5 .2 11.2 4.2 15.2s9.7 5.6 15.2 4.2l60.1-15c5.6-1.4 10.8-4.3 14.9-8.4L576.1 358.7l-71-71L375.9 417z"
          />
        </svg>
        traveler123
      </p>
    </div>
    <div
      aria-label="Review text"
      class="_reviewInfoBox_1b1c0f"
      role="contentinfo"
    >
      <p
        aria-expanded="false"
        class="_reviewText_1b1c0f"
        id="review-text-Amazing Experience"
      >
        This experience was absolutely wonderful and unforgettable. The journey was packed with excitement a...
      </p>
      <button
        aria-controls="review-text-Amazing Experience"
        aria-expanded="false"
        aria-label="Expand review text"
        class="_readMoreButton_1b1c0f"
      >
        Read More
      </button>
    </div>
  </article>
</DocumentFragment>
`;
